protocol-concepts/DynamicFees.md

```diff
@@ -0,0 +1,382 @@
+- Change on a per-swap basis
+- Allow for any fee percentage (e.g., 4.9 bips, 10 bips)
+- Be updated at various intervals (yearly, per block, or per transaction)
+
+## Motivation and Benefits of Dynamic Fees
+
+1. **Improved Pricing of Volatility:** Adapt fees to market volatility, similar to traditional exchanges adjusting bid-ask spreads.
+2. **Order Flow Discrimination:** Price different types of trades (e.g., arbitrage vs. uninformed) more accurately.
+3. **Improved Market Efficiency and Stability:** Fees can adjust to reflect real-time market conditions, optimizing for both liquidity providers and traders. Dynamic fees could help dampen extreme market movements by adjusting incentives in real-time.
+4. **Enhanced Capital Efficiency and Liquidity Provider Returns:** By optimizing fees, pools can attract more liquidity and facilitate more efficient trading. More accurate fee pricing could lead to better returns for liquidity providers, potentially attracting more capital to pools.
+5. **Better Risk Management:** During high volatility, fees can increase to protect liquidity providers from impermanent loss.
+6. **Customizable Strategies:** Enable complex fee strategies for specific token pairs or market segments.
+
+## Dynamic Fees Use Cases
+
+1. **Volatility-Based Fees:** Adjust fees based on the historical or realized volatility of the asset pair.
+2. **Volume-Based Fees:** Lower fees during high-volume periods to attract more trades, and increase fees during low-volume periods to compensate liquidity providers.
+3. **Time-Based Fees:** Implement different fee structures for different times of day or days of the week, based on historical trading patterns.
+4. **Market Depth-Based Fees:** Adjust fees based on the current liquidity depth in the pool.
+5. **Cross-Pool Arbitrage Mitigation:** Dynamically adjust fees to discourage harmful arbitrage between different pools or exchanges.
+6. **Gas Price-Responsive Fees:** Adjust fees based on network congestion and gas prices to ensure profitability for liquidity providers.
+7. **Event-Driven Fees:** Implement special fee structures during significant market events or token-specific occurrences.
+8. **Lookback approach:** Set the fee to match the most profitable fee tier of external pools with the same asset pair over a recent period.
+9. **Price oracle approach:** Use an external price oracle to determine the correct asset price and adjust fees based on how trades move the pool price relative to this external price.
+10. **Price momentum approach:** Analyze recent price history and asymmetrically adjust fees based on trade direction.
+11. **Asset composition approach:** Lower fees for trades that balance the pool and higher fees for trades that imbalance it.
+12. **Transaction-source based approach:** Provide lower fees for transactions routed through certain aggregators or sources less likely to be arbitrage trades.
+
+## Dynamic Fees Mechanism
```

Comment:
here i think its worth briefly mentioning that Uniswap v4 supports two ways of updating the fee:

* periodically with a call to PoolManager
* on every swap, by returning the fee from beforeSwap


then we can link to the guide

Reply:
we can also mention that whether or not a pool has a dynamic fee is specified at pool-creation and its immutable